keps/sig-node/kubelet-resource-metrics-endpoint.md

```diff
@@ -0,0 +1,208 @@
+* Monitoring Metrics: Metrics for observability and introspection of the cluster, which are used by end-users, operators, devs, etc. 
+
+
+The Kubelet’s [JSON Summary API](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/stats/v1alpha1/types.go) is currently used by the [Metrics Server](https://github.com/kubernetes-incubator/metrics-server#kubernetes-metrics-server).  It contains far more metrics than are required by the Metrics Server.
+
+[Prometheus](https://prometheus.io/) is commonly used for exposing metrics for kubernetes components, and the [Prometheus Operator](https://github.com/coreos/prometheus-operator#prometheus-operator), which Sig-Instrumentation works on, is commonly used to deploy and manage metrics collection.
+
+[OpenMetrics](https://openmetrics.io/) is a new prometheus-based metric standard which supports both text and protobuf.  
+
+[GRPC](https://grpc.io/) is commonly used for interfaces between components in kubernetes, such as the [Container Runtime Interface](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto).  GRPC uses [protocol-buffers](https://developers.google.com/protocol-buffers/docs/overview) (protobuf) for serialization and deserialization, which is more performant than other formats.
+
+## Motivation
+
+The Kubelet Summary API is a source of both Resource and Monitoring Metrics.  Because of it’s dual purpose, it does a poor job of both.  It provides much more information than required by the Metrics Server, as demonstrated by [kubernetes/kubernetes#68841](https://github.com/kubernetes/kubernetes/pull/68841).  Additionally, we have pushed back on adding metrics to the Summary API for monitoring, such as DiskIO or tcp/udp metrics, because they are expensive to collect, and not required by all users.
+
+This proposal deals with the first problem, which is that the Summary API is a poor provider of Resource Metrics.  It proposes a purpose-built API for supplying Resource Metrics.
+
+### Goals
+
+* [Primary] Provide the minimum set of metrics required to serve the Resource Metrics API
+* [Secondary] Minimize the CPU and Memory footprint of the metrics server due to collecting metrics
+  * Perform efficiently at frequent (sub-second) rates of metrics collection
+* [Secondary] Use a format that is familiar to the kubernetes community, which can be consumed by common monitoring pipelines, and is interoperable with commonly-used monitoring pipelines.
+
+### Non-Goals
+
+* Deprecate or remove the Summary API
+* Add new Resource Metrics to the metrics server (e.g. Ephemeral Storage)
+* Detail how the kubelet will collect metrics to support this API.
+* Determine what the pipeline for “Monitoring” metrics will look like
+
+## Proposal
+
+The kubelet will expose an endpoint at `/metrics/resource/v1alpha1` in prometheus text exposition format using the prometheus client library.
+
+
+### API
+
+```
+Name: container_cpu_usage_core_nanoseconds_total
```

Comment:
it's kind-of expected that time-based metrics are exposed in seconds (as per standard Prometheus conventions).  Any reason not to do that?

Reply:
Would changing it to `container_cpu_usage_nanocore_seconds_total` work?  I don't want to have to round to whole seconds of cpu usage...

Reply:
spoke offline.  I forgot prometheus metrics are floats... Changing to `container_cpu_usage_seconds_total`